src/AccessibilityInsights.SharedUx/Controls/HierarchyControl.xaml.cs

```diff
@@ -847,11 +847,8 @@ public void FileBug(HierarchyNodeViewModel vm = null)
             else
             {
                 // File a new bug
-                Logger.PublishTelemetryEvent(TelemetryAction.Scan_File_Bug, new Dictionary<TelemetryProperty, string>
-                {
-                    { TelemetryProperty.By, FileBugRequestSource.Hierarchy.ToString() },
-                    { TelemetryProperty.IsAlreadyLoggedIn, IssueReporter.IsConnected.ToString(CultureInfo.InvariantCulture) },
-                });
+                var telemetryEvent = TelemetryEventCreator.ForIssueFilingRequest(FileBugRequestSource.Hierarchy);
```

Comment:
Some questions for reviewers:
- does it make sense for the TelemetryAction to be separate from the event creator (as in this PR), or should the event creator return a `TelemetryEvent` that contains both the action and the properties? 
- what is the right return type for the event creator?

Reply:
Having a `TelemetryEvent` class makes total sense, and since we build the same thing in 3 places (lots of copy & paste there), having a single method to create the `TelemetryEvent` also makes sense. If we create a Factory class, we should name it as such, and this method would probably be `CreateFileIssueTelemetryEvent(FileBugRequestSource source)` or something similar. The calling code would just be `Logger.PublishTelemetryEvent(CreateFileIssueTelemetryEvent(FileBugRequestSource.Hierarachy));` You'd obviously need to create a `Logger.PublishTelemetryEvent(TelemetryEvent event)` method, but that's pretty trivial.

That said, the following bounds come to mind:
1. This model only makes sense where we have duplicated code (i.e., variations of the same event are being published from multiple places)
2. The `TelemetryEvent` class has a natural affinity to the `Logger` class and should be in the same namespace
3. The code that builds the `TelemetryEvent` object has a natural affinity to the input data of the method, and should be in the same namespace as those classes.
4. The corollary of points 2 and 3 is that you are likely to eventually end up with multiple Factories--some may be a full class, while others may just be a method within a class

---
In reply to: [277115343](https://github.com/Microsoft/accessibility-insights-windows/pull/328#discussion_r277115343) [](ancestors = 277115343)

Reply:
Thanks Dave. The first bound you list is partly why I'm not as sure about bundling the action & properties together. Maybe evaluating more examples (I'm thinking of the beaker button, for instance) would help me think about it further. I'll address the other suggestions, think about this some more, & publish the PR for review early next week.